### DISCUSSION TITLE
Using cargo crates in main.rs

### DISCUSSION THREAD

**Discussion by mehmedaltug** [tauri-apps/tauri#9460] [Q&A] [answered]
I have a simple project in tauri, using yarn and vanilla html css js. I wanted to use a library from crates.io so i added it to cargo.toml inside the src-tauri directory but when i have that dependency tauri fails to build. So i searched the web but couldnt find anything about this topic. 

My question is:
What is the proper way of using crates from crates.io(for rust obv.) if it is even possible?

**Comment by FabianLars** [comment]
> I wanted to use a library from crates.io so i added it to cargo.toml inside the src-tauri directory but when i have that dependency tauri fails to build

This is generally the correct approach. Using crates in a tauri app works the same as in every rust app, this includes potential conflicts between crates, especially those that compile (to) C code internally.

Mind sharing the error so we can help you here?

**Reply by mehmedaltug** [reply]
So upon further testing I saw that adding the crate doesnt break the program, problem is using it, and it is probably about library locations.

Error message is quite long so i wrote it to a file:
[file link](https://raw.githubusercontent.com/mdev896/err_msg/main/README.md)

And the simplified version:
```yarn run v1.22.22
warning ../../package.json: No license field
$ tauri dev
        Info Watching /home/mdev/Desktop/tauri-app/src-tauri for changes...
   Compiling tauri-app v0.0.0 (/home/mdev/Desktop/tauri-app/src-tauri)
warning: unused import: enigo::*`
` --> src/main.rs:4:5
  |
4 | use enigo::*;`
 ` |     ^^^^^^^^
  | `
  ` = note: #[warn(unused_imports)] on by default `

`error: linking with cc failed: exit status: 1 `
   ` | (...) ` 
` = note: /usr/bin/ld: cannot find -lxdo: No such file or directory
          collect2: error: ld returned 1 exit status
  `        

`warning: tauri-app (bin "tauri-app") generated 1 warning
error: could not compile tauri-app (bin "tauri-app") due to 1 previous error; 1 warning emitted
```

~~I think it can be solved by adding the location of the crate i am using(or external crates location) to the list maybe?~~

**EDIT**: Forgot to mention that I created a new project for the output message.

**UPDATE**: It was about the crate I was using, used a different crate for testing and it worked? So this thread is crate specific, the crate "enigo" [https://crates.io/crates/enigo](https://crates.io/crates/enigo)

**Reply by FabianLars** [reply, marked as answer]
This error means that you missed this part: https://github.com/enigo-rs/enigo?tab=readme-ov-file#runtime-dependencies

**Reply by mehmedaltug** [reply]
Silly me, I was sure I installed required packages but after checking them, turns out I forgot them. 

Thank you, and I am really sorry if I took your time.
